packages/bulk-edit/src/browser/bulk-edit-contribution.ts

```diff
@@ -0,0 +1,106 @@
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable } from 'inversify';
+import { Widget } from '@theia/core/lib/browser/widgets/widget';
+import { CommandRegistry } from '@theia/core/lib/common';
+import { AbstractViewContribution } from '@theia/core/lib/browser/shell/view-contribution';
+import { BulkEditCommands } from './bulk-edit-commands';
+import { MonacoBulkEditService } from '@theia/monaco/lib/browser/monaco-bulk-edit-service';
+import { TabBarToolbarContribution, TabBarToolbarRegistry } from '@theia/core/src/browser/shell/tab-bar-toolbar';
+import { BulkEditTreeWidget, BULK_EDIT_TREE_WIDGET_ID } from './bulk-edit-tree';
+
+@injectable()
+export class BulkEditContribution extends AbstractViewContribution<BulkEditTreeWidget> implements TabBarToolbarContribution {
+    private workspaceEdit: monaco.languages.WorkspaceEdit;
+
+    constructor(private readonly bulkEditService: MonacoBulkEditService) {
+        super({
+            widgetId: BULK_EDIT_TREE_WIDGET_ID,
+            widgetName: 'Refactor Preview',
+            defaultWidgetOptions: {
+                area: 'bottom'
+            }
+        });
+        this.bulkEditService.setPreviewHandler((edits: monaco.languages.WorkspaceEdit) => this._previewEdit(edits));
+    }
+
+    registerCommands(registry: CommandRegistry): void {
+        super.registerCommands(registry);
+        registry.registerCommand(BulkEditCommands.APPLY, {
+            isEnabled: widget => this.withWidget(widget, () => true),
+            isVisible: widget => this.withWidget(widget, () => true),
+            execute: widget => this.withWidget(widget, () => this.apply())
+        });
+        registry.registerCommand(BulkEditCommands.DISCARD, {
+            isEnabled: widget => this.withWidget(widget, () => true),
+            isVisible: widget => this.withWidget(widget, () => true),
+            execute: widget => this.withWidget(widget, () => this.discard())
+        });
+    }
+
+    async registerToolbarItems(toolbarRegistry: TabBarToolbarRegistry): Promise<void> {
+        toolbarRegistry.registerItem({
+            id: BulkEditCommands.APPLY.id,
+            command: BulkEditCommands.APPLY.id,
+            tooltip: 'Apply Refactoring',
+            priority: 0,
+        });
+        toolbarRegistry.registerItem({
+            id: BulkEditCommands.DISCARD.id,
+            command: BulkEditCommands.DISCARD.id,
+            tooltip: 'Discard Refactoring',
+            priority: 1,
+        });
+    }
+
+    protected withWidget<T>(widget: Widget | undefined = this.tryGetWidget(), cb: (bulkEdit: BulkEditTreeWidget) => T): T | false {
+        if (widget instanceof BulkEditTreeWidget) {
+            return cb(widget);
+        }
+        return false;
+    }
+
+    private async _previewEdit(workspaceEdit: monaco.languages.WorkspaceEdit): Promise<monaco.languages.WorkspaceEdit> {
```

Comment:
```suggestion
    private async previewEdit(workspaceEdit: monaco.languages.WorkspaceEdit): Promise<monaco.languages.WorkspaceEdit> {
```
https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#no_underscore_private

Reply:
Fixed